### PR TITLE
Document R package and add R CMD Check via {tic} on Travis

### LINKED ISSUE #701: R: Document all function arguments and add examples

### LINKED ISSUE #704: R: CI
Travis or GHA? -> TBD

@hannesmuehleisen 
Any important secrets that we need to be aware of?

### PR DESCRIPTION
**Pull request by pat-s**
fixes #701 
fixes #704

@hannesmuehleisen 

I added `FIXME` placeholders for all arguments that still need to be documented.
Do you want to do this? You know the arguments and their scope better than me :)

On the side, R CMD Check is now running: https://travis-ci.org/github/cwida/duckdb/jobs/704308816

I see two things:

- The `lib/` directory is massive with 61 MB (locally on macOS it is only 9 MB)
- We have a length1 condition error in a function: https://travis-ci.org/github/cwida/duckdb/jobs/704308816#L1039
- We should discuss how to deal with `src/duckdb.hpp`: https://travis-ci.org/github/cwida/duckdb/jobs/704308816#L981. Maybe @krlmlr has an idea?

### CODE REVIEW

**Comment by hannesmuehleisen** [comment]
did the fixmes

File: tools/rpkg/R/Connection.R [hunk 1]
```diff
@@ -1,13 +1,3 @@
-#' @title DuckDB Connection
-#'
-#' @description
-#' TBD
-#'
-#' @param duckdb_driver FIXME
-#' @param debug FIXME
-#'
-#' @name duckdb_connection
-#' @export
```
Comment (pat-s) [inline tools/rpkg/R/Connection.R]: Was this intended?

File: tools/rpkg/R/Result.R [hunk 2]
```diff
@@ -1,10 +1,3 @@
-#' @title DuckDB Result Set
-#' @description  TBD
-#'
-#' @param connection FIXME
-#' @param stmt_lst FIXME
-#' @export
-#' @name duckdb_result
```
Comment (pat-s) [inline tools/rpkg/R/Result.R]: Was this intended?

File: tools/rpkg/R/Result.R [hunk 3]
```diff
@@ -20,8 +13,6 @@ duckdb_result <- function(connection, stmt_lst) {
   return(res)
 }
 
-#' @rdname duckdb_result
-#' @export
```
Comment (pat-s) [inline tools/rpkg/R/Result.R]: Was this intended?

File: tools/rpkg/R/Result.R [hunk 4]
```diff
@@ -32,8 +23,6 @@ duckdb_execute <- function(res) {
   }
 }
 
-#' @rdname duckdb_result
-#' @export
```
Comment (pat-s) [inline tools/rpkg/R/Result.R]: Was this intended?

Comment (hannesmuehleisen) [reply tools/rpkg/R/Result.R]: the constructor does not need to be exported?

File: tools/rpkg/R/Connection.R (hunk 1 above)
Comment (hannesmuehleisen) [reply tools/rpkg/R/Connection.R]: the constructor duckdb_connection should not be exported

File: tools/rpkg/R/Result.R (hunk 2 above)
Comment (pat-s) [reply tools/rpkg/R/Result.R]: This results in 

```r
checking for missing documentation entries ... WARNING
  Undocumented S4 methods:
    generic 'dbBind' and siglist 'duckdb_result'
    generic 'dbClearResult' and siglist 'duckdb_result'
    generic 'dbColumnInfo' and siglist 'duckdb_result'
    generic 'dbFetch' and siglist 'duckdb_result'
    generic 'dbGetInfo' and siglist 'duckdb_result'
    generic 'dbGetRowCount' and siglist 'duckdb_result'
    generic 'dbGetRowsAffected' and siglist 'duckdb_result'
    generic 'dbGetStatement' and siglist 'duckdb_result'
    generic 'dbHasCompleted' and siglist 'duckdb_result'
    generic 'dbIsValid' and siglist 'duckdb_result'
    generic 'show' and siglist 'duckdb_result'
```

`duckdb_result()` needs to be exported and documented.

I did so now. You might want to document the two arguments of `duckdb_result()` and have a look at the failing tests. These look like true positives on the first look but I might be wrong.

**Comment by krlmlr** [comment]
I tweaked a bit, now creating `.h` files. Do we need to fix `extension/parquet/parquet_amalgamation.py` to respect the extension of the header file?

**Comment by hannesmuehleisen** [comment]
I like the sed solution

**Comment by krlmlr** [comment]
`sed` won't work on macOS though, due to different semantics of the `-i` switch on BSD. `-i''` might work.

**Comment by krlmlr** [comment]
@pat-s: Can you please take a look at the `sed` failure, see https://travis-ci.org/github/cwida/duckdb/jobs/704564173#L490.

File: .travis.yml [hunk 5]
```diff
@@ -576,29 +576,38 @@ matrix:
 
     - os: linux
       name: R Package
-      if: (type = push AND branch = master) OR type = pull_request OR tag =~ /^v\d+\.\d+\.\d+$/
+      if: (type = push) OR type = pull_request OR tag =~ /^v\d+\.\d+\.\d+$/
```
Comment (hannesmuehleisen) [inline .travis.yml:579]: any reason the master check has disappeared?

**Review by hannes — APPROVED** [review]

**Comment by hannesmuehleisen** [comment]
passes now on all platforms. merge? @pat-s

File: .travis.yml (hunk 5 above)
Comment (krlmlr) [reply .travis.yml:579]: The net effect is that R is now also checked for every branch, even if it doesn't have a pull request yet. However, R is lined up last and needs to wait anyway. Maybe we can shift it up in the build matrix, or disable altogether?

Comment (hannesmuehleisen) [reply .travis.yml:579]: I'm okay running it all the time but then just delete the if line altogether

Comment (pat-s) [reply .travis.yml:579]: This was meant to be a temporary change, just to get started with the R build in a branch without a PR. One R build for non-PR builds might be ok, the full suite can only be run on master then?

**Comment by hannesmuehleisen** [comment]
So what exactly is the benefit of using this tic package? I see a lot of pain getting it setup since its not yet on CRAN and it just seems like yet another dependency.

**Comment by krlmlr** [comment]
tic is a build-time dependency that will make the CI process much easier, at the very least during development.

I wonder why the R builds worked before on Windows.

**Comment by hannesmuehleisen** [comment]
> tic is a build-time dependency that will make the CI process much easier, at the very least during development.
Hm.

> I wonder why the R builds worked before on Windows.
I think it has to do with new rtools. There seems to be a cygwin DLL conflict now. I am working on it.

**Comment by hannesmuehleisen** [comment]
Ran into a different issue while investigating this: The DuckDB amalgamation (duckdb.cpp) fails to compile on Windows with MinGW (used by Rtools) in 32 Bit mode because the compiler runs out of memory (I read something about a 2 GB limit somewhere). So we would first need to change the amalgamation script to create multiple source files before the compilation & check could pass.

**Comment by krlmlr** [comment]
Do we really care about Windows i386? We could put x64 it into `SystemRequirements` and be done.

**Comment by hannesmuehleisen** [comment]
> Do we really care about Windows i386? We could put x64 it into `SystemRequirements` and be done.

I have heard that many places in developing countries still run i386 windows. We would like to help them as well of course. But I think we have a battle plan. Will try to implement tomorrow.

**Comment by pat-s** [comment]
> So what exactly is the benefit of using this tic package? I see a lot of pain getting it setup since its not yet on CRAN and it just seems like yet another dependency.

It is a DSL for CI for the R language. CRAN or not does not really make a difference here.
It automates many important CI steps in a convenient way (dep installation, R CMD Check, codecov, deployment (optional)). 
We use it in all of our pkgs but we are also biased here since we are the devs 😄 

If you absolutely dislike it, we can remove it again.

## rtools

We should not mix R 4.0 and rtools 3.5.
I was starting to use R4.0 in my last commit to get the most recent R version rather than v3.
I see you reverted to rtools 3.5 - is it worth looking at all the failed builds? Not sure what might conflict here (according to your comment), this is not my field of expertise :)

Should we proceed with the R3.x approach on Windows for now? And also tackle the 32-bit issue separately to move on?

**Comment by hannesmuehleisen** [comment]
Given time constraints I'm okay requiring 64 bit for now for the R package.

**Comment by hannesmuehleisen** [comment]
And no, go back to rtools 4 of course it was just a test.

**Comment by hannesmuehleisen** [comment]
I'm still sceptical wrt the tic package, but defer to your judgement on this for now. Curious to see where it is in a year.

**Comment by krlmlr** [comment]
An example how to declare a 64-bit OS: https://github.com/eddelbuettel/rcppcctz/blob/master/DESCRIPTION#L17.

For 32-bit Windows we might want to change amalgamation to create one file per directory, or perhaps to copy the original code and adapt `src/Makevars`?

**Comment by pat-s** [comment]
Ich denke wir sind einen Schritt weiter: https://travis-ci.org/github/cwida/duckdb/jobs/706221096#L1123.

Hat jemand direkt ne Idee dazu?

**Comment by krlmlr** [comment]
I'll test the run on a Windows machine.

**Comment by hannesmuehleisen** [comment]
Yes I suspect this to be related to tic.

**Comment by pat-s** [comment]
> Yes I suspect this to be related to tic.

{tic} does not change anything in the normal build-check routine. 

I've also run the build-check part without {tic} to demonstrate that again: https://travis-ci.org/github/cwida/duckdb/jobs/707341741

I am out of ideas. For most time, I thought its a 32-bit issue (maybe it still is) but using `--no-multiarch` does not help.
I cannot reproduce the issue in my Windows VM either.

The only similar issue with respect to R packages is found in {stringi} but [without a solution](https://github.com/gagolews/stringi/issues/313) unfortunately.

Handing over to @krlmlr for the moment.

**Comment by pat-s** [comment]
@hannesmuehleisen I suggest merging this PR and tackling the windows issues on Travis separately in #755.

**Comment by hannesmuehleisen** [comment]
@pat-s the tests are still not passing. Why would you suggest I merge it then?

**Comment by pat-s** [comment]
Superseeded by multiple smaller PRs.

**Pull request closed without merging**